### PR TITLE
chore: refactor bigtable tracers

### PR DESCRIPTION
**Pull request by mutianf**
Refactor ApiTracers used in the client:

- `BigtableTracer` will record bigtable specific metrics. 
- `CompositeTracer` now contains a list of all the tracers and a separate list for `BigtableTracer`s and will call the list of `BigtableTracer`s for Bigtable specific metrics.
- `MetricsTracer` now extends `BigtableTracer`.
- Removed `HeaderTracer` and implementing it in `MetricsTracer` and moved extracting values from server-timing headers to `metrics/Util`

Customers should be able to extend `BigtableTracer` for their own implementations.

### CODE REVIEW

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BigtableTracer.java [hunk 1]
```diff
@@ -18,155 +18,22 @@
...
-  }
-
-  @Override
-  public void batchRequestSent(long elementCount, long requestSize) {
-    for (ApiTracer child : children) {
-      child.batchRequestSent(elementCount, requestSize);
-    }
-  }
+/** A Bigtable specific {@link ApiTracer} that includes additional contexts. */
+public abstract class BigtableTracer extends BaseApiTracer {
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BigtableTracer.java:24]: I would mark this BetaApi

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/CompositeTracerFactory.java [hunk 2]
```diff
@@ -24,15 +23,11 @@
...
 import java.util.List;
 
-/**
- * A Bigtable specific {@link ApiTracerFactory} that combines multiple {@link ApiTracerFactory} into
- * a single one.
- */
-@InternalApi("For internal use only")
-public class BigtableTracerFactory extends BaseApiTracerFactory {
+/** Combines multiple {@link ApiTracerFactory} into a single {@link ApiTracerFactory}. */
+public class CompositeTracerFactory extends BaseApiTracerFactory {
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/CompositeTracerFactory.java:27]: Why remove internalapi?

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/Util.java [hunk 3]
```diff
@@ -95,10 +101,23 @@ static TagValue extractStatus(Future<?> future) {
         ATTEMPT_EPOCH_KEY.name(),
         Arrays.asList(String.valueOf(Instant.EPOCH.until(Instant.now(), ChronoUnit.MICROS))));
     // This should always be true
-    if (apiCallContext.getTracer() instanceof BigtableTracer) {
-      int attemptCount = ((BigtableTracer) apiCallContext.getTracer()).getAttempt();
+    if (apiCallContext.getTracer() instanceof CompositeTracer) {
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/Util.java:104]: shouldnt this be instanceof BigtableTracer?

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/CompositeTracer.java [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+import com.google.common.collect.ImmutableList;
+import java.util.ArrayList;
+import java.util.List;
+import javax.annotation.Nullable;
+import org.threeten.bp.Duration;
+
+/**
+ * Combines multiple {@link ApiTracer}s and {@link BigtableTracer}s into a single {@link ApiTracer}.
+ */
+public class CompositeTracer extends BigtableTracer {
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/CompositeTracer.java:28]: Internal?

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/HeaderTracerStreamingCallable.java [hunk 5]
```diff
@@ -72,20 +67,17 @@ public void call(
 
   private class HeaderTracerResponseObserver<ResponseT> implements ResponseObserver<ResponseT> {
 
+    private BigtableTracer tracer;
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/HeaderTracerStreamingCallable.java:70]: final

Comment (igorbernstein2) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/HeaderTracerStreamingCallable.java:70]: I think the other fields should be as well

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/CompositeTracer.java (hunk 4 above)
Comment (mutianf) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/CompositeTracer.java:28]: Ah this class shouldn't be public :)

**Review by igorbernstein2 — APPROVED** [review]
lgtm

**Pull request merged**